src/ledger/LedgerTxn.cpp

```diff
@@ -192,7 +280,10 @@ populateLoadedEntries(KeySetT const& keys,
 
     for (auto const& key : keys)
     {
-        if (res.find(key) == res.end())
+        // If the key was not loaded (but not due to metering), we should put
+        // a nullptr entry in the result.
+        if (res.find(key) == res.end() &&
+            (!lkMeter || !lkMeter->loadFailed(key)))
```

Comment:
We don't win anything by doing this check, do we? The key will be loaded (again) during the transaction application, only then fail. If we have loaded the entry, we should always cache it as to not waste the work  that has already be done. We should only skip prefetching keys that don't fit into quota before the entry was loaded.

Reply:
This is so that we can disambiguate non-existent entries (which result in a nullptr in the cache) from entries which were not loaded due to insufficient quota. We are cacheing everything that we do load, up until we realize that the quota has been exceeded.

Reply:
`updateReadQuotasForKey` should only be called for entries (because as I've mentioned keys are not counted towards the read bytes). Thus we'd only run out of quota after we have loaded the entry. That's not a huge deal, but OTOH if we remove this code, then I'm not sure if we lose anything, but we'd definitely have less code to maintain.

Reply:
If I remove this code, then we will have key/nullptr in the cache for key/entry pairs that are not loaded due to insufficient quota (but do exist in the bucketlist). This is the desired behavior for non-existent entries, but we don't want that when the entry exists but just wasn't funded, right?

Reply:
As I've cleared up my understanding of how `loadKeysWithLimits`, I think this is mostly fine. One scenario where this is somewhat sub-optimal is when `lkMeter->updateReadQuotasForKey(*currKeyIt, entrySize);` happens to exceed quota, i.e. we would load the entry, but still record that the load has failed. But I think that's a rare enough scenario to not worry much about it.

Reply:
> somewhat sub-optimal is when lkMeter->updateReadQuotasForKey(*currKeyIt, entrySize); happens to exceed quota

Yeah that's not great, I intend to forgo that unfunded load with a subsequent xdr level optimization to partially deserialize the entry and confirm its within quota before reading the entire entry in a follow up pr. :)